openedx/openedx-platform
# Save user state for XBlocks in Blockstore-based content libraries

This PR introduces the changes needed for XBlocks in Blockstore to save their user state into CSM.

Before this PR, all student state for Blockstore blocks was ephemeral (in-process dict store).

## Notes / questions:
- The main risk factor of this PR is that it adds non-course keys to the `course_id` field in CSM. If any code (like analytics?) reads course keys directly out of CSM and doesn't have graceful handling for key types it doesn't recognize, it could cause an issue. With the included changes to opaque-keys, calling `CourseKey.from_string(...)` on these values will raise `InvalidKeyError` since they're not `CourseKey`s. (But calling `LearningContextKey.from_string(...)` will work for both course and library keys.)
- This PR introduces a slight regression for the Studio view of XBlocks in Blockstore content libraries: their state is now lost from request to request. I have a follow up PR to give them a proper studio-appropriate state store, but I want to review it separately so it doesn't hold up this PR and we can test this PR on its own.
- Is there a [performance?] reason why CSM has no validation to check internal consistency, like an assert that the `course_id` matches the `usage_key.course_id` and the `module_type` matches the `usage_key.block_type`?

## Test Instructions:

* Run Blockstore (`cd blockstore; make easyserver`)
* Install the very latest commit of [Ramshackle](https://github.com/open-craft/ramshackle/)
* Log in to Studio and the LMS
* Go to http://localhost:18010/ramshackle/ and view the new "Learn" tab for any XBlock in a Blockstore-based content library. Confirm it saves state to CSM:
    <img width="810" alt="Screen Shot 2019-09-11 at 9 38 58 AM" src="https://user-images.githubusercontent.com/945577/64717067-6c522e80-d478-11e9-94e0-6617646c721d.png">
